### PR TITLE
[DOC] Generate citable boilerplates for workflows

### LINKED ISSUE #192: Generate citable boilerplates
**Is your feature request related to a problem? Please describe.**
This feature request is not related to a problem, _per se_. I believe that users would benefit from an autogenerated methods section with citations that they could copy, in its entirety, into their manuscripts.

**Describe the solution you'd like**
Ultimately, I think it would be best to leverage NiBetaSeries' Nipype-based implementation to incorporate node-wise or workflow-wise descriptions like fMRIPrep, but that would be a lot of work. It might be more feasible to have a formattable multiline string in `cli/run.py` that is updated based on the parameters used. That's a bit more like how we do it for our workflows in NiMARE (see [here](https://github.com/neurostuff/NiMARE/blob/e17629fb0793a593a9b244ca4fe9dbbc3412f592/nimare/workflows/macm.py#L31) for an example).

**Describe alternatives you've considered**
A section on the RTD site (like fMRIPrep used to have) might be a good substitute, but wouldn't be customized based on the specific parameters used. I also don't know if BIDS derivatives json files are designed to contain full methods sections (i.e., if #98 can deal with this without adding a new output).

**Additional context**
I think that this feature would be particularly useful after your JOSS article is published, so you can ensure that the article is cited as people use the tool.

### PR DESCRIPTION
**Pull request by tsalo**
## Summary

Fixes #192. An example boilerplate (run locally) for an LSA run: [NiBetaSeries citation boilerplate.pdf](https://github.com/HBClab/NiBetaSeries/files/3610917/NiBetaSeries.citation.boilerplate.pdf).

This is still a work in progress. FS is assumed to be supported here, so this has to wait until #204 is dealt with. I also need to update the dependencies and tests. The actual text could probably use some TLC too. 

## List of changes proposed in this PR (pull-request)

- Use Niworkflows' LiterateWorkflow class for workflows instead of the Nipype one.
- Add configurable descriptions to workflows.
- Add tools from fMRIPrep for compiling descriptions into a boilerplate and converting the boilerplate to a number of output formats (tex, md, and html).
- Move BibTeX file originally just used for website. Now used for website _and_ boilerplates.

### CODE REVIEW

File: src/nibetaseries/workflows/model.py [hunk 1]
```diff
@@ -75,7 +78,97 @@ def init_betaseries_wf(name="betaseries_wf",
...
+autocorrelation.
+
+After fitting each model, the parameter estimate map associated with the
+target trial's regressor was retained and concatenated into a 4D image with all
+other trials from that condition, resulting in a set of N 4D images of varying
+sizes, where N refers to the number of conditions in the task.
+""".format(nistats_ver=nistats_ver, smooth_str=smooth_str, hrf=hrf_model,
+           confound_str=confound_str, hpf=high_pass)
+    elif estimator == 'lss' and hrf_model == 'fir':
+        fir_delays = [0, 1, 2, 3, 4]
```
Comment (tsalo) [inline src/nibetaseries/workflows/model.py]: TODO: Remove when #204 (or similar) is merged.

**Comment by jdkent** [comment]
Just wanted to say amazing work :100: , and I will take a closer look tomorrow.

File: src/nibetaseries/workflows/model.py [hunk 2]
```diff
@@ -75,7 +78,97 @@ def init_betaseries_wf(name="betaseries_wf",
...
+regressors. In a finite impulse response model, the BOLD response for each of a
+set of delays following the onset of the trial is modeled as an impulse, which
+can be used to capture the hemodynamic response model associated with that
+condition. In the models applied in this workflow, delay-specific FIR
+regressors corresponding to {delays} volumes after the target trial were
+included, while all other trials from that condition were modeled in a second
+set of FIR regressors, and other conditions were modeled in their own sets of
+FIR regressors.
+In addition to condition regressors, {confound_str}a
+high-pass filter of {hpf} Hz (implemented using a cosine drift model) were
```
Comment (jdkent) [inline src/nibetaseries/workflows/model.py]: if `{confound_str}` is empty, `were` should become `was`.

File: src/nibetaseries/data/references.bib [hunk 3]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+year = 2014
+}
+@article{gramfort2014mne,
+  title={MNE software for processing MEG and EEG data},
+  author={Gramfort, Alexandre and Luessi, Martin and Larson, Eric and Engemann, Denis A and Strohmeier, Daniel and Brodbeck, Christian and Parkkonen, Lauri and H{\"a}m{\"a}l{\"a}inen, Matti S},
+  journal={Neuroimage},
+  volume={86},
+  pages={446--460},
+  year={2014},
+  publisher={Elsevier}
```
Comment (jdkent) [inline src/nibetaseries/data/references.bib]: add 
```
doi = {10.1016/j.neuroimage.2013.10.027}
```

File: src/nibetaseries/data/references.bib [hunk 4]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+}
+@article{gramfort2013meg,
+  title={MEG and EEG data analysis with MNE-Python},
+  author={Gramfort, Alexandre and Luessi, Martin and Larson, Eric and Engemann, Denis A and Strohmeier, Daniel and Brodbeck, Christian and Goj, Roman and Jas, Mainak and Brooks, Teon and Parkkonen, Lauri and others},
+  journal={Frontiers in neuroscience},
+  volume={7},
+  pages={267},
+  year={2013},
+  publisher={Frontiers}
```
Comment (jdkent) [inline src/nibetaseries/data/references.bib]: add:
```
doi = {10.3389/fnins.2013.00267}
```

File: src/nibetaseries/data/references.bib [hunk 5]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+title = {Nipype},
+year = 2018,
+doi = {10.5281/zenodo.596855},
+publisher = {Zenodo},
+journal = {Software}
+}
+@article{fisher1915frequency,
```
[the remark this answers is not in the document]
Comment (jdkent) [reply src/nibetaseries/data/references.bib]: add space before and after equals sign for the fields.

File: src/nibetaseries/data/references.bib [hunk 6]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+issn = {1662-5196},
+journal = {Frontiers in Neuroinformatics},
+language = {English},
+title = {Machine learning for neuroimaging with scikit-learn},
+url = {https://www.frontiersin.org/articles/10.3389/fninf.2014.00014/full},
+volume = 8,
+year = 2014
+}
+@article{gramfort2014mne,
+  title={MNE software for processing MEG and EEG data},
```
Comment (jdkent) [inline src/nibetaseries/data/references.bib]: add a space before and after the equals sign for these fields

File: src/nibetaseries/data/references.bib [hunk 7]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+  author={Gramfort, Alexandre and Luessi, Martin and Larson, Eric and Engemann, Denis A and Strohmeier, Daniel and Brodbeck, Christian and Parkkonen, Lauri and H{\"a}m{\"a}l{\"a}inen, Matti S},
+  journal={Neuroimage},
+  volume={86},
+  pages={446--460},
+  year={2014},
+  publisher={Elsevier}
+}
+@article{gramfort2013meg,
+  title={MEG and EEG data analysis with MNE-Python},
```
Comment (jdkent) [inline src/nibetaseries/data/references.bib]: add a space before and after the equals sign for these fields.

File: src/nibetaseries/data/references.bib [hunk 8]
```diff
@@ -130,3 +130,118 @@ @article{Verstynen2014
...
+  author={Gramfort, Alexandre and Luessi, Martin and Larson, Eric and Engemann, Denis A and Strohmeier, Daniel and Brodbeck, Christian and Goj, Roman and Jas, Mainak and Brooks, Teon and Parkkonen, Lauri and others},
+  journal={Frontiers in neuroscience},
+  volume={7},
+  pages={267},
+  year={2013},
+  publisher={Frontiers}
+}
+@Article{Hunter2007,
+  Author    = {Hunter, J. D.},
```
Comment (jdkent) [inline src/nibetaseries/data/references.bib]: modify spaces so there is one space before and after the equals sign.

**Comment by tsalo** [comment]
@jdkent I think that I've address your review now.

One thing that I'm having trouble with is ensuring that the Zenodo-based references match up to the versions specified in the Dockerfile. For the NiBetaSeries reference, at least, it might be good to include a line about updating `references.bib` to match the current release on Zenodo to your release checklist.

**Comment by jdkent** [comment]
Thanks @tsalo 

> For the NiBetaSeries reference, at least, it might be good to include a line about updating references.bib to match the current release on Zenodo to your release checklist.

I'm a little confused by this point, the zenodo DOIs you are using for NiBetaSeries and Nipype are the general dois that reference all/any versions of the software, so those references would not need to change based on the versions of the software that are installed.

Sorry, I may have missed the point you are trying to make.

File: src/nibetaseries/workflows/model.py (hunk 2 above)
Comment (jdkent) [reply src/nibetaseries/workflows/model.py]: still need this solution:
```
is_mult_confs='were' if len(confound_str) else 'was')
```
(clever btw)

File: src/nibetaseries/workflows/model.py [hunk 9]
```diff
@@ -75,7 +78,98 @@ def init_betaseries_wf(name="betaseries_wf",
...
+
+A least squares- all (LSA) model was generated following the method described in
+@Rissman2004, using Nistats {nistats_ver}.
+Prior to modeling, preprocessed data were {smooth_str}masked,
+and mean-scaled over time.
+Preprocessed data were subjected to a general linear model in which each trial
+was modeled in its own regressor. Each trial-specific regressor was convolved
+with a "{hrf}" hemodynamic response function for the model.
+In addition to condition regressors, {confound_str}a
+high-pass filter of {hpf} Hz (implemented using a cosine drift model) were
```
Comment (jdkent) [inline src/nibetaseries/workflows/model.py]: still need this solution:
```
is_mult_confs='were' if len(confound_str) else 'was')
```
here

**Comment by tsalo** [comment]
> I'm a little confused by this point, the zenodo DOIs you are using for NiBetaSeries and Nipype are the general dois that reference all/any versions of the software, so those references would not need to change based on the versions of the software that are installed.
> 
> Sorry, I may have missed the point you are trying to make.

I think it would be nice to cite the Zenodo DOI associated with the version currently being used. Citation tracking fro Zenodo DOIs isn't great at the moment, but in the future hopefully that won't be the case, and it will be possible to track version-specific usage.

That said, this doesn't seem to be a solved issue anywhere. I haven't seen a solid automated solution anywhere. In tedana, we just update the citation in our docs. We don't do anything about it in our boilerplate. As such, maybe it's not worth trying to do in this PR.

**Comment by jdkent** [comment]
Thanks for clarifying! I'll sleep on a solution, but that may be out of the scope for this PR at least.

File: src/nibetaseries/workflows/model.py [hunk 10]
```diff
@@ -79,7 +82,99 @@ def init_betaseries_wf(name="betaseries_wf",
...
+In addition to condition regressors, {confound_str}a
+high-pass filter of {hpf} Hz (implemented using a cosine drift model) {is_mult_confs}
+included in the model.
+AR(1) prewhitening was applied in each model to account for temporal
+autocorrelation.
+
+After fitting each model, the parameter estimate map associated with the
+target trial's regressor was retained and concatenated into a 4D image with all
+other trials from that condition, resulting in a set of N 4D images of varying
+sizes, where N refers to the number of conditions in the task.
```
[the remark this answers is not in the document]
Comment (tsalo) [reply src/nibetaseries/workflows/model.py]: Makes sense to me. I'll push the change to all three descriptions.

**Comment by tsalo** [comment]
I think that the review has been addressed. Is this good to go?

**Comment by jdkent** [comment]
Sorry @tsalo! The only thing holding me back on this is seeing what a generated report looks like now.

**Comment by tsalo** [comment]
I had to change how the final workflow description in model was being formatted (apparently you have to dedent first, then add the substrings), but now it looks good:

image

I also added a `--boilerplate` argument to make it easier to just get the boilerplate without running the workflow.

**Comment by PeerHerholz** [comment]
@tsalo re getting the boilerplate without running the workflow: e.g., if a user already run the workflow but forget to include the `--boilerplate` argument or want to regenerate it?

**Comment by tsalo** [comment]
@PeerHerholz The boilerplate will automatically be generated no matter what, so I think the main benefit to the new option is getting the writeup before running on real data, as in for a preregistration or just to better understand what _will_ happen when you do have real data to run on.

**Comment by PeerHerholz** [comment]
fantastic! Just because I was running into this problem lately: what if a certain software's version changes between generating the boilerplate and actually running the analysis? I know, it might be rather unlikely, but can happen.

**Comment by tsalo** [comment]
That's definitely a good point, and I hope that the majority of people are grabbing the boilerplate from their actual run. Although, I do generally assume that folks are running BIDS Apps from Docker or Singularity images, which would reduce the likelihood of accidentally changing versions.

**Comment by PeerHerholz** [comment]
True that and I totally agree with you. I was just thinking about it based on some BIDS Apps I'm currently writing and seeing inexperienced users updating images mid-analysis. Do you think a pointer (as done for [fmriprep](https://fmriprep.readthedocs.io/en/stable/faq.html#a-new-version-of-fmriprep-has-been-published-when-should-i-upgrade)) would be too much?

**Comment by tsalo** [comment]
I think it's worth adding to the documentation. There's currently only one FAQ so if nothing else it would make that page feel more complete.

**Comment by jdkent** [comment]
Excellent points @tsalo and @PeerHerholz! I'll open an issue for adding "when should I update" to the FAQ.

**Pull request merged**